include/zephyr/bluetooth/l2cap.h

```diff
@@ -366,6 +377,48 @@ struct bt_l2cap_chan_ops {
 	 *  @param chan The channel which was reconfigured
 	 */
 	void (*reconfigured)(struct bt_l2cap_chan *chan);
+
+#ifdef CONFIG_BT_L2CAP_SEG_RECV
+	/** @brief Handle L2CAP segments directly
+	 *
+	 *  This is an alternative to @ref bt_l2cap_chan_ops.recv. They cannot
+	 *  be used together.
+	 *
+	 *  This is called immediately for each received segment.
+	 *
+	 *  Unlike with @ref bt_l2cap_chan_ops.recv, flow control is explicit.
+	 *  Each time this handler is invoked, the remote has permanently used
+	 *  up one credit. Use @ref bt_l2cap_chan_give_credits to give credits.
+	 *
+	 *  The start of a SDU is marked by `seg_offset == 0`. The end of a SDU
+	 *  is marked by `seg_offset + seg_len == sdu_len`.
+	 *
+	 *  The stack guarantees that..
+	 *    - The sender had the credit.
+	 *    - The SDU length does not exceed MTU. `sdu_len <= chan.rx.mtu`.
+	 *    - The segment length does not exceed MPS. `seg_len <=
+	 *      chan.rx.mps`.
+	 *    - Segments are consecutive. `(seg_offset[i] + seg_len[i] ==
+	 *      seg_offset[i + 1]) || seg_offset[i + 1] == 0`.
+	 *    - The previous SDU is completed before a new one is started.
+	 *      `(seg_offset[i] + seg_len[i] == sdu_len[i]) || seg_offset[i + 1]
+	 *      != 0`.
+	 *    - There is no trailing data in the last segment of an SDU.
+	 *      `seg_offset[i] + seg_len[i] <= sdu_len[i]`.
+	 *
+	 *  @note With this alternative API, the application is responsible for
+	 *  setting the RX MTU and MPS. The MPS must not exceed @ref
+	 *  BT_L2CAP_RX_MTU.
+	 *
+	 *  @param chan The receiving channel.
+	 *  @param sdu_len Byte length of the SDU this segment is part of.
+	 *  @param seg_offset The byte offset of this segment in the SDU.
+	 *  @param seg_len The byte length of this segment.
+	 *  @param seg The segment payload. Size is `seg_len`.
+	 */
+	void (*seg_recv)(struct bt_l2cap_chan *chan, uint16_t sdu_len,
+			 uint16_t seg_offset, uint16_t seg_len, uint8_t *seg);
```

Comment:
Would it maybe be better to use`net_buf_simple` instead of `seg_len` and `seg`. You can then make the call to the callback as:
```c
chan->chan.ops->seg_recv(&chan->chan, chan->_sdu_len, seg_offset, &seg->b);
```

That might be friendlier to applications who  may want to take advantage of net_buf parsing APIs for processing the segment payload.

Or, if you don't think net_buf_simple is helpful to applications, at least fix the types: it should be `size_t` and `const void *` instead of `uint8_t` and `uint8_t *`.

Another thing I'd question is the types for `sdu_len` and `seg_offset`. Normally we've tried to use logical types rather than fixed-width types in public APIs, so `size_t` and `off_t` in this case.

Reply:
I generally concur with the use of the logical types. One just have to remember that we need to add value range verification then (e.g. if using `size_t` for a `uint16_t` value or if using the (signed) `off_t` for a `uint16_t`).